core/trino-main/src/main/java/io/trino/cost/TopNRankingStatsRule.java

```diff
@@ -0,0 +1,201 @@
+            double distinctCount = estimateCorrelatedDistinctCount(node.getOrderingScheme().orderBy(), sourceStats, INDEPENDENCE_FACTOR);
+            // This should be greater than 1.0, otherwise we will overestimate the rank distinct count.
+            double rowsPerDistinctValue = ceil(maxRowsPerPartition / distinctCount);
+            rankDistinctCount = ceil(node.getMaxRankingPerPartition() / rowsPerDistinctValue);
+            rowsPerPartition = min(maxRowsPerPartition, rankDistinctCount * rowsPerDistinctValue);
+        }
+        else {
+            // In case of DENSE_RANK, each unique row gets a unique rank. The rank is incremental.
+            // For example, ordering by ROW1, ROW2:
+            // ROW1   ROW2   DENSE_RANK
+            // A      A      1
+            // A      A      1
+            // B      B      2
+            // B      B      2
+            // C      C      3
+            // C      C      3
+            // In the above example, the rank is incremental. Therefore, the maximum number of rows per partition can be
+            // at most (maxRank * rowsPerDistinctValue). In the above case rowsPerDistinctValue is 2 based on distinct
+            // count. Hence, the number of rows per partition for max rank 2 should be 4. Whereas, the distinct count
+            // for rank symbol should be 2 i.e. maxRank.
+            double distinctCount = estimateCorrelatedDistinctCount(node.getOrderingScheme().orderBy(), sourceStats, INDEPENDENCE_FACTOR);
+            // This should be greater than 1.0, otherwise we will underestimate the rows per partition.
+            double rowsPerDistinctValue = ceil(maxRowsPerPartition / distinctCount);
+            rowsPerPartition = min(maxRowsPerPartition, node.getMaxRankingPerPartition() * rowsPerDistinctValue);
+            rankDistinctCount = min(distinctCount, node.getMaxRankingPerPartition());
+        }
+
+        if (isNaN(rowsPerPartition)) {
+            rowsPerPartition = maxRowsPerPartition;
+        }
+
+        if (isNaN(rankDistinctCount)) {
+            rankDistinctCount = rowsPerPartition;
+        }
+
+        PlanNodeStatsEstimate.Builder estimateBuilder = PlanNodeStatsEstimate.buildFrom(sourceStats);
+        fixOrderBySymbolDistinctCount(
+                node.getOrderingScheme().orderBy(),
+                partitionCount,
+                rowsPerPartition,
+                maxRowsPerPartition,
+                sourceStats,
+                estimateBuilder);
+
+        double outputRowsCount = partitionCount * rowsPerPartition;
+        return Optional.of(estimateBuilder
+                .setOutputRowCount(outputRowsCount)
+                .addSymbolStatistics(node.getRankingSymbol(), SymbolStatsEstimate.builder()
+                        // Note: if we assume no skew, we could also estimate highValue
+                        // (as rowsPerPartition), but underestimation of highValue may have
```

Comment:
Do real query actually work without hi stats? Usually it's something like:
```
select rank() over (partitioned bv x, y order by z desc) as r WHERE r < 1
```
Hence I believe there will be a `FilterNode(r < 1)` above that won't be estimated without lo/hi. Consider adding an e2e plan test (I don't recall if there a class already with such tests).

> we could also estimate highValue (as rowsPerPartition)

we probably could do that if otherwise we don't have stats at all